### ISSUE TITLE
Query, PathVariable and BodyParameter support

### ISSUE THREAD

**Issue by FelixResch** [Web25/http#6]
The project should support parsing of the URL Query, PathVariables like @doompickaxe mentioned and direct parsing of BodyParameters if the are UrlFormEncoded.

They should be made available in the IncomingHttpRequest and OutgoingHttpResponse classes.

Maybe we could add support for template requests like this:

```kotlin
val http = Http()
val prepared = http.prepare().get("http://example.org/users/{uid}")
val user1 = prepared.bind("uid", 1).response() //also executes .execute()
val user2 = prepared.bind("uid", 2).response()
```

- [x] Query parsing
- [x] PathVariable
- [ ] BodyParameters (same API for client and server, discard current client API)
- [x] PreparedRequests?

**Comment by FelixResch** [comment]
Here a few API examples of this change:

# Query Parsing

Every request URI in HTTP can have a query. This is the part after the `?` in an URI. 

To access the query parameters of `/rest/users/12?format=xml` the API could be restructured to be used like this

```kotlin
handler {request, response -> 
    val format = request.query["format"]
}
```

The same goes for client requests. So creating a request to the above URI could look like this

```kotlin
val request = http.get("http://example.org/")
request.query["format"] = "xml"
request.execute()
```

# Path Variables

Like @doompickaxe mentioned the use of dynamic paths with path variables is necessary to create REST APIs with this library. 

To achieve this, the path of a request could be changed to a new class `HttpPath` that handles variables and is able to create a "real" path string.

The API could look like this (and would also make prepared requests possible):

```kotlin
// Server: Added Path: /users/{uid} | Sent Path: /users/1
handler { request, response -> 
    val uid = request.path["uid"].toInt()
    ...
}

//Client: Prepared Requests
val request = http.get("http://example.org/{uid}")
request.path["uid"] = 1
val userData1 = request.response().reponseBody()
request.path["uid"] = 2
val userData2 = request.response().reponseBody()
```

The HttpPath class could look like this (no method bodies):

```kotlin
fun HttpPath(val path: String) {
    operator fun set(key: String, value: [String, Number, ...])
    operator fun get(key: String): String
    fun toPath(): String // creates the real path
    override fun toString() = toPath()
}
```

# Body Parameters

This change would require to change the internals of http requests to have an object for http entities that extend HttpEntity. To not break the API, the methods that already exist map directly to the entity object.

The interface could look like this:

```kotlin
interface HttpEntity {
    fun getBytes(): ByteArray
    fun asString() = String(getBytes())
}
```

When sending a URLFormEncoded Entity an instance of URLFormEncodedHttpEntity can be created and key value pairs added to it. Same goes for JSON, YAML, XML and other Entity formats.

```kotlin
val entity = URLFormEncodedHttpEntity()
entity["uid"] = 1
entity["username"] = "test"
val response = http.post("http://example.org/users/").entity(entity).response()
```

On the server the right Decoder needs to be selected based on the `Content-Type` header of the request and the Data needs to be decoded. The handling is then performed by a normal handler that needs to check which data type the entity is and can access it based on the data type.

```kotlin
handler {request, response -> 
    if(request.entity is SomeDataClass) {  // serializer found to deserialize the data

    } else if (request.entity is URLFormEncodedHttpEntity) {  // classic key value pairs of data

    } else {
        // FAIL because entity is in the wrong format
    }
}
```

## Serializers

The HttpContext should keep track of serializers that can serialize objects to data and data to objects. If a serializer can't deserialize a data because it can't determine the class it should deserialize to, helpers should add information on how the data should be processed.

```kotlin
val http = Http()
val context = http.context
context.add(GsonSerializer())
context.add(YamlSerializer())
context.add(PathInformation("/some/path", SomeDataClass::class))
...
```

**Comment by doompickaxe** [comment]
# Query Parsing
I think query is the wrong word. There are GET and POST parameters, so we need to differentiate these

```Kotlin
val request = http.get("http://example.org/")
request.getParameters["format"] = "xml"
request.execute() 
```

or

```Kotlin
handler {request, response -> 
    val format = request.postParameters["name"]
}
```

**Comment by FelixResch** [comment]
# Query Parsing

Maybe query is misleading in this context, but according to [RFC 3986, Section 3.4](https://tools.ietf.org/html/rfc3986#section-3.4) is the correct term for the "GET parameters"

The naming of *GET* and *POST* parameters has it's origin in PHP where *GET* parameters are part of the URI query sent to the server (defined in RFC 3986, 3.4) and *POST* parameters is URLFormEncoded data sent by an HTML form.

I would recommend using standards compliant naming so switching from other platforms is easier.

**Comment by doompickaxe** [comment]
Maybe we can rewrite the way cookies are implemented in the way I implemented query parsing.

```kotlin
val request = http.get("http://example.com")
request.cookies["oreo"] = "yummy"
request.execute()
```

**Comment by FelixResch** [comment]
Could be done using an object. Doing so would obsolete #15 because we wouldn't need the map if we used an object.

```kotlin
val cookies = object {
    private val map = mutableMapOf<String, HttpCookie>()
    operator fun get(name: String): HttpCookie {
        val cookie = map[name]
        if(cookie == null)
            throw HttpCookieException("Cookie $name not found")
        return cookie
    }
}
```

Of course also adding setters and other functions necessary for traversing all cookies. I will give it a try and create a new branch with this API and see how well it works.

**Comment by FelixResch** [comment]
Moving this idea to #16 because it could also affect other parts of the API

**Comment by FelixResch** [comment]
@doompickaxe would you mind if I took care of the BodyParameters/Entities?

**Comment by doompickaxe** [comment]
No, i just saw i forgot to implement query parsing for response, but I will do it immediately

**Comment by doompickaxe** [comment]
@FelixResch I just realised, if you implemented #17 , we also should do that for query and headers to be consistent. Maybe there is a more general solution.

**Comment by FelixResch** [comment]
Currently working on the headers (see #16) but we can also add the query. I will merge the cookie stuff into the main development branch for a clearer structure.

**Comment by doompickaxe** [comment]
# Prepared Requests
Binding on path variables is a great idea, but what do you think about binding on GET parameters @FelixResch ?
Should we make something like this possible?

```kotlin
val http = Http()
val request = http.prepare("http://example.com/{action}").bindPathVar("action", "show")
val user1 = request.bindPathVar.bindGet("uid", 1).response()
val user2 = request.bindPathVar.bindGet("uid", 1).response()
```

**Comment by FelixResch** [comment]
You mean as part of the query? So performing `user1` would result in a call to `/show?user=1` and a calling `user2` as `/show?user=2`?.

Maybe it's me, but i fail to see how this would improve the API as we could simply create a method that would set a query parameter and call that instead of a own bind method that does the same. 

But we could add a list of required query parameters for a prepared request, so requests that need certain parameters will fail if not all parameters are present. 

An API could maybe look like this:

```kotlin
val users = http.prepare("http://example.com/{action}/{uid}")
    .requireQuery(name = "format", default = "json", possible = arrayOf("xml", "yaml", "json")) // for the required query parameter, including some checks
    .require(QueryParameter(name = "format", default = "json", possible = arrayOf("xml", "yaml", "json"))) //Alternative approach for above call

val userFail = users.bind("action", "show").response()    //would fail, because not all parameters are set
val userWrongFormat = users.bind("action", "show").bind("uid", 1).bind("format", "txt").response()    // would also fail, because validation would not check out

val user1 = users.bind("action", "show").bind("uid", 1).response() // would result in a call to /show/1?format=json and the returned json
```

Personally I would prefer the second approach on required parameter setting as PathVariables (or parameters) could also be part of the prepared request and could be checked more easily with proper internal design.

**Comment by FelixResch** [comment]
# HttpPath

We could, instead of using one string and a list, create a class that encapsulates both and offers a simpler and cleaner API. This class would be constructed using a string and parses the individual parts into path elements/components/parts/... . It also has methods to get & set path variables, so accessing them would be easier.

This would change the API from 

```kotlin
val request = http.get("http://example.org/{action}")
request.pathVars["action"] = "show"
```

to 

```kotlin
val request = http.get("http://example.org/{action}")
request.path["action"] = "show"
```

This might seem like a little API change but, looks much cleaner, considering header, cookies and the query are now accessed like this.

## Getting the Request Path

The class would also need a method to create the path that is actually going to be used, so the request can be performed. This method should throw an exception if a required path component has not been added yet, so no invalid requests can be sent.

## Parsing the Request Path

The server needs to be able to parse the request path of an incoming request if it has path variables. Therefore the class must know the base path, that has been called to match the path variables. To keep matching simple we should define a fixed set of characters that delimits path components from each other. When looking at the structure of a path the following characters could be used as delimiters:

* `/`
* `.`

Others could be added later if needed. 

Also incoming path variables should be immutable as all handlers should be able to see the path variables.

## Summary

TL;DR: We need a class that

* can split the path of a request into multiple components
* those parameters can be
    * static - their value never changes
    * variables - their value can be changed
* keep track of variables
* match variable markers of a parsed path with values in an unmarked path sent to the request
* construct a path based on values from a map and a parsed path and throw an exception if component values are missing

**Comment by doompickaxe** [comment]
What's new here? Anything left?

**Comment by FelixResch** [comment]
not really, I think a merging session is required, but afterwards we can close this issue

**Comment by doompickaxe** [comment]
HttpQuery is a part of HttpPath, so this should be linked. Then this issue can be closed
